pallets/stateful-storage/src/tests.rs

```diff
@@ -0,0 +1,2533 @@
+			StatefulChildTree::<<Test as Config>::KeyHasher>::try_read(
+				&msa_id,
+				PALLET_STORAGE_PREFIX,
+				ITEMIZED_STORAGE_PREFIX,
+				&key,
+			)
+			.unwrap()
+			.unwrap();
+		let previous_hash = previous_page.get_hash();
+
+		let payload = ItemizedSignaturePayload {
+			actions: BoundedVec::try_from(actions).unwrap(),
+			target_hash: previous_hash,
+			msa_id,
+			expiration: 10,
+			schema_id,
+		};
+		let encode_data_new_key_data = wrap_binary_data(payload.encode());
+		let owner_signature: MultiSignature = pair.sign(&encode_data_new_key_data).into();
+
+		// act
+		assert_err!(
+			StatefulStoragePallet::apply_item_actions_with_signature(
+				RuntimeOrigin::signed(caller_1),
+				delegator_key.into(),
+				owner_signature,
+				payload
+			),
+			Error::<Test>::CorruptedState
+		)
+	});
+}
+
+#[test]
+fn upsert_page_with_signature_having_page_id_out_of_bounds_should_fail() {
+	new_test_ext().execute_with(|| {
+		// arrange
+		let caller_1 = test_public(1);
+		let (msa_id, pair) = get_signature_account();
+		let delegator_key = pair.public();
+		let schema_id = UNDELEGATED_PAGINATED_SCHEMA;
+		let page_id = (<Test as Config>::MaxPaginatedPageId::get() + 1).into();
+		let payload = generate_payload_bytes::<PaginatedPageSize>(Some(100));
+		let payload = PaginatedUpsertSignaturePayload {
+			payload,
+			target_hash: PageHash::default(),
+			msa_id,
+			expiration: 10,
+			schema_id,
+			page_id,
+		};
+		let encode_data_new_key_data = wrap_binary_data(payload.encode());
+		let owner_signature: MultiSignature = pair.sign(&encode_data_new_key_data).into();
+
+		// act
+		assert_err!(
+			StatefulStoragePallet::upsert_page_with_signature(
+				RuntimeOrigin::signed(caller_1),
+				delegator_key.into(),
+				owner_signature,
+				payload
+			),
+			Error::<Test>::PageIdExceedsMaxAllowed
+		)
+	})
+}
+
+#[test]
+fn upsert_page_with_signature_having_expired_payload_should_fail() {
+	new_test_ext().execute_with(|| {
+		// arrange
+		let caller_1 = test_public(1);
+		let (msa_id, pair) = get_signature_account();
+		let delegator_key = pair.public();
+		let schema_id = UNDELEGATED_PAGINATED_SCHEMA;
+		let page_id = 1;
+		let block_number = 10;
+		let payload = generate_payload_bytes::<PaginatedPageSize>(Some(100));
+		let payload = PaginatedUpsertSignaturePayload {
+			payload,
+			target_hash: PageHash::default(),
+			msa_id,
+			expiration: block_number,
+			schema_id,
+			page_id,
+		};
+		let encode_data_new_key_data = wrap_binary_data(payload.encode());
+		let owner_signature: MultiSignature = pair.sign(&encode_data_new_key_data).into();
+
+		// act
+		System::set_block_number(block_number);
+		assert_err!(
+			StatefulStoragePallet::upsert_page_with_signature(
+				RuntimeOrigin::signed(caller_1),
+				delegator_key.into(),
+				owner_signature,
+				payload
+			),
+			Error::<Test>::ProofHasExpired
+		)
+	})
+}
+
+#[test]
+fn upsert_page_with_signature_having_out_of_window_payload_should_fail() {
```

Comment:
I don't see a specific test for a replay. That might be a good one to have.

Reply:
I have a draft PR up with the SignedExtrinsic stuff, that will need to be merged before we do this final merge to `main` (unless we just want to merge this & I'll rebase my PR off of `main`). I need to do a bit of test cleanup/rework in there anyway, I can add one for replay.

Reply:
1 step replay is kinda tested with different hash values. If it has the same hash value it means the page didn't change.